File: LeetCode/DataStruct/JS/Leetcode952-largest-component-size-by-common-factor.js
var largestComponentSize = function (nums) {
    let [n, res] = [nums.length, 1];
    let [fa, cnt] = [new Array(n).fill(0), new Array(n).fill(1)];
    for (let i = 0; i < n; i++) {
        fa[i] = i;
    }
    let app = new Map();
    for (let i = 0; i < n; i++) {
        const x = nums[i];
        for (let j = 1; j <= Math.floor(x / j); j++) {
            if (x % j == 0) {
                if (j > 1) {
                    if (!app.get(j)) app.set(j, [i]);
                    else app.get(j).push(i);
                }
                if (j == Math.floor(x / j)) continue;
                if (!app.get(Math.floor(x / j))) app.set(Math.floor(x / j), [i]);
                else app.get(Math.floor(x / j)).push(i);
            }
        }
    }
    let find = (x) => {
        if (fa[x] != x) fa[x] = find(fa[x]);
        return fa[x];
    }

    for (let v of app.values()) {
        for (let i = 1; i < v.length; i++) {
            let [a, b] = [v[0], v[i]];
            if (find(a) != find(b)) {
                cnt[find(a)] += cnt[find(b)];
                fa[find(b)] = find(a);
                res = Math.max(res, cnt[find(a)]);
            }
        }
    }
    return res;
};

File: LeetCode/DataStruct/Rust/Leetcode952-largest-component-size-by-common-factor.rs
use std::collections::HashMap;

pub struct UnionFind {
  id: Vec<usize>,
  sizes: Vec<usize>,
}

impl UnionFind {
  pub fn new(n: usize) -> Self {
    let mut id = vec![0; n];
    let mut sizes = vec![0; n];
    for i in 0..n {
      id[i] = i;
      sizes[i] = 1;
    }
    Self { id, sizes }
  }

  pub fn find(&mut self, x: usize) -> usize {
    if self.id[x] != x {
      self.id[x] = self.find(self.id[x]);
    }
    self.id[x]
  }

  pub fn union(&mut self, a: usize, b: usize) -> usize {
    let (x, y) = (self.find(a), self.find(b));
    if x != y {
      self.id[x] = y;
      self.sizes[y] += self.sizes[x];
    }
    self.sizes[y]
  }
}

impl Solution {
  pub fn breakdown(a: i32) -> Vec<i32> {
    let mut res = vec![];
    let mut n = a;
    let mut i = 2;
    while i * i <= n {
      if n % i == 0 {
        // 重复的只添加一次
        res.push(i);
        while n % i == 0 {
          n /= i;
        }
      }
      i += 1;
    }

    if n != 1 {
      res.push(n);
    }
    res
  }

  pub fn largest_component_size(nums: Vec<i32>) -> i32 {
    let n = nums.len();
    // 质因数对应数字的索引
    let mut m = HashMap::new();
    for (i, a) in nums.iter().enumerate() {
      for b in Self::breakdown(*a) {
        let e = m.entry(b).or_insert_with(Vec::new);
        e.push(i)
      }
    }

    let mut res = 1;
    let mut uf = UnionFind::new(n);
    for values in m.into_values() {
      for i in 1..values.len() {
        let size = uf.union(values[i - 1], values[i]);
        res = res.max(size);
      }
    }
    res as i32
  }
}
